### PR TITLE
x11-libs/wxGTK: multilib support

### PR DESCRIPTION
**Pull request by nE0sIghT**
Multilib support for wxGTK 2.8 and 3.0 as i requested here: https://bugs.gentoo.org/show_bug.cgi?id=510708
This changes needed for games-emulation/pcsx2: https://bugs.gentoo.org/show_bug.cgi?id=510710

I added multilib support to eselect-wxwidgets via epatch here. This change available also at https://github.com/nE0sIghT/eselect/commits/eselect-wxwidgets-multilib and must be applied to eselect main repository

### CODE REVIEW

**Comment by mgorny** [comment]
@axs-gentoo, could you take a look at this one, please? :)

**Comment by mgorny** [comment]
@leio!

File: app-eselect/eselect-wxwidgets/files/0001-wxwidgets.eselect-multilib-support.patch [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+ 
+ DESCRIPTION="Manage the system default wxWidgets profile"
+ MAINTAINER="wxwidgets@gentoo.org"
+-VERSION="20140423"
++VERSION="20150329"
+ 
+ WXCONFFILE="${EROOT}"/var/lib/wxwidgets/current
+-WXCONFDIR="${EROOT}"/usr/lib/wx/config
++WXCONFDIR_64="${EROOT}"/usr/lib/wx/config
++WXCONFDIR_32="${EROOT}"/usr/lib32/wx/config
```
Comment (mgorny) [inline app-eselect/eselect-wxwidgets/files/0001-wxwidgets.eselect-multilib-support.patch]: No, no, no. There's some eselect library for multilib support. If you're doing that, you need to support all ABIs, not just x86+amd64.

File: x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	fi
+
+	econf ${myconf}
+}
+
+multilib_src_compile() {
+	emake
+
+	if [[ -d contrib/src ]]; then
+		cd contrib/src
```
Comment (mgorny) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: ||die

File: x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	default
+
+	if [[ -d contrib/src ]]; then
+		cd contrib/src
+		emake DESTDIR="${D}" install
+	fi
+}
+
+multilib_src_install_all() {
+	cd "${S}"/docs
```
Comment (mgorny) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: ||die :)

File: x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	dodoc changes.txt readme.txt todo30.txt
+	newdoc base/readme.txt base_readme.txt
+	newdoc gtk/readme.txt gtk_readme.txt
+
+	if use doc; then
+		dohtml -r "${S}"/docs/html/*
```
Comment (mgorny) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: dohtml is deprecated, use `dodoc -r "${S}"/docs/html`

**Comment by nE0sIghT** [comment]
Pushed fixes.
Branch eselect-wxwidgets-multilib at https://github.com/nE0sIghT/eselect overwritten

**Comment by nE0sIghT** [comment]
> And anyway, I'm not sure if it's a good idea to allow combining different wxwidgets versions for different ABIs. The module would certainly be simpler if you just set the same thing for all libdirs (this is e.g. how eselect-opengl does it).

So maybe drop eselect-wxwidgets multilib changes?
It can be used only for out-of-portage builds (and maybe for broken portage builds, but i don't think there is exists such) that relies on wxconfig utility in $PATH.

**Comment by mgorny** [comment]
That's some option too. Especially that the current eselect module is broken for me anyway.

**Comment by axs-gentoo** [comment]
Still need anything from me here?  It seems all issues have been resolved now.  I second the dropping of multilib changes to eselect-wxwidgets.

**Comment by mgorny** [comment]
Yes, I only looked at the eselect module :P. I would really appreciate if you could take a look at this one, so I could focus on the repo-mirror-ci scripting.

**Comment by mgorny** [comment]
Sorry for replying this late. I don't really like the way get_libdir is hacked in there. Could you by any chance look into replacing the sed with something better? Possibly an upstreamable patch?

**Comment by nE0sIghT** [comment]
I replaced sed fix with configure patch in last commit.
Patch for wxGTK-2.8 is huge because of whitespace/quotes changes in regenerated configure script... Do not know how to avoid this (i used same version of autoconf - 2.59)

**Comment by nE0sIghT** [comment]
Upstream PR: https://github.com/wxWidgets/wxWidgets/pull/59

**Comment by nE0sIghT** [comment]
[Upstream said](https://github.com/wxWidgets/wxWidgets/pull/59#issuecomment-127772039) that libdir problems should not affect 3.0 wxGTK.
I tested x32 chroot and it compiled well (and linked - as i see) without any sed/patch. However i can not [test ia64](https://bugs.gentoo.org/show_bug.cgi?id=499984), but that bug was reported against 2.8 version also.

**Comment by mgorny** [comment]
Well, I personally think all that magic should be killed with fire. If possible, please try to backport the upstream change to 2.8. That would certainly be much cleaner. After that, it should be possible to build wxGTK out-of-source.

**Comment by nE0sIghT** [comment]
I [tried to backport](https://github.com/nE0sIghT/wxWidgets/commits/feature/multilib) library search code and was able to pass configure stage, but now i hit strange relocation/-fPIC problem...

**Comment by nE0sIghT** [comment]
Problem was in autoconf.eclass - it uses autoconf-2.69 when WANT_AUTOCONF=2.5 specified.
I pushed commit with backported patch - is it ok to depend and directly call autoconf-2.59?

**Comment by mgorny** [comment]
Nope, that's not allowed. I think you need to figure out how to make it happy about new autoconf.

**Comment by nE0sIghT** [comment]
I added autoconf patch and rebased PR.
Tested with ABI_X86="32 64 x32"

File: x11-libs/wxGTK/files/wxGTK-2.8.12.1-bakefile.patch [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/build/aclocal/bakefile.m4 b/build/aclocal/bakefile.m4
+index 88478ca..aa04b16 100644
+--- a/build/aclocal/bakefile.m4
++++ b/build/aclocal/bakefile.m4
+@@ -837,7 +837,7 @@ AC_DEFUN([AC_BAKEFILE],
+     AC_SUBST(OBJCXXFLAGS)
+ 
+ 
+-    BAKEFILE_BAKEFILE_M4_VERSION="0.2.8"
++    BAKEFILE_BAKEFILE_M4_VERSION="0.2.9"
```
Comment (mgorny) [inline x11-libs/wxGTK/files/wxGTK-2.8.12.1-bakefile.patch:19]: Just to be clear, does this really matter?

**Comment by mgorny** [comment]
Great work! I'm going to test it now, and hopefully commit it soon enough :).

Do you want to become wxGTK co-maintainer by any chance? You seem to be handling this creepy thing great, and wxwidgets team would certainly use some help.

File: x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild [hunk 6]
```diff
@@ -0,0 +1,189 @@
+# Copyright 1999-2015 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Header: /var/cvsroot/gentoo-x86/x11-libs/wxGTK/wxGTK-2.8.12.1-r1.ebuild,v 1.13 2015/03/31 20:05:20 ulm Exp $
+
+EAPI="5"
+WANT_AUTOCONF=2.5
```
Comment (mgorny) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: Do you actually need this anymore?

**Comment by mgorny** [comment]
Bad news (2.8), 32&64:

```
checking for Xinerama... no
configure: WARNING: Xinerama not found; disabling wxDisplay
checking for -lSM - X11 session management... no
configure: WARNING: libSM not found; disabling session management detection
```

I think you need to backport a bit more :).

File: x11-libs/wxGTK/files/wxGTK-2.8.12.1-bakefile.patch (hunk 5 above)
Comment (nE0sIghT) [inline x11-libs/wxGTK/files/wxGTK-2.8.12.1-bakefile.patch:19]: Yes. Without that change build failed with

```
configure: error: Versions of Bakefile used to generate makefiles (0.2.9) and configure (0.2.8) do not match.
```

File: x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild (hunk 6 above)
Comment (nE0sIghT) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: I added this because with next autoconf release build may break again

Comment (mgorny) [inline x11-libs/wxGTK/wxGTK-2.8.12.1-r2.ebuild]: It will be a build-time breakage, so better than forcing dep that people will forget to update :).

**Comment by mgorny** [comment]
You also want to `MULTILIB_CHOST_TOOLS` wx-config-3.0.

**Comment by nE0sIghT** [comment]
I updated backported patch for Xinerama and SM and removed autoconf dependency.
Also added MULTILIB_CHOST_TOOLS part, but it has turned a bit tricky because wx-config tools already laid in libdir.

> Do you want to become wxGTK co-maintainer by any chance?

I will be glad to help if i can :-)

**Comment by mgorny** [comment]
Ouch. Let's revert that chost tool part then, I see that the eclass uses /usr/lib\* directly anyway. I'll just revert it myself while committing. Now I'm going to do some live surgery on my partitions, then I'll get back to you and test 2.8 again.

**Comment by mgorny** [comment]
```
+*wxGTK-3.0.2.0-r2 (06 Aug 2015)
+*wxGTK-2.8.12.1-r2 (06 Aug 2015)
+
+  06 Aug 2015; Michał Górny <mgorny@gentoo.org>
+  +files/wxGTK-2.8.12.1-autoconf.patch, +files/wxGTK-2.8.12.1-bakefile.patch,
+  +files/wxGTK-2.8.12.1-libdir.patch, +wxGTK-2.8.12.1-r2.ebuild,
+  +wxGTK-3.0.2.0-r2.ebuild:
+  Enable multilib support. Backport some build system patches.
+  https://github.com/gentoo/gentoo-portage-rsync-mirror/pull/142 by nE0sIghT.
```

I've also pinged wxGTK maintainers about you. I'll let you know when I get a reply.

**Pull request closed without merging**